src/cla/tcp.rs

```diff
@@ -0,0 +1,511 @@
+                       if let TcpClPacket::SessTerm(_) = packet {
+                           //breaks loop, tasks finished, dropped sender, connection closed
+                           break;
+                       }
+                    }
+                }
+                _ = interval.tick() => {
+                    self.send_packet(&TcpClPacket::KeepAlive).await;
+                }
+            }
+        }
+    }
+    async fn send_packet(&mut self, packet: &TcpClPacket) {
+        if let Err(err) = packet.serialize(&mut self.stream_writer).await {
+            error!("Error while serializing packet: {}", err);
+        }
+        if let Err(err) = self.stream_writer.flush().await {
+            error!("Error while flashing tcp sending queue: {}", err);
+        }
+    }
+}
+
+struct Connection {
+    stream: TcpStream,
+    addr: SocketAddr,
+}
+
+impl Connection {
+    async fn negotiate_session(&mut self) -> anyhow::Result<(SessInitData, SessInitData)> {
+        let mut sess_init_data = SessInitData {
+            keepalive: 30,
+            segment_mru: 64000,
+            transfer_mru: 64000,
+            node_id: "test2".into(),
+        };
+        let session_init = TcpClPacket::SessInit(sess_init_data.clone());
+        session_init.serialize(&mut self.stream).await?;
+        self.stream.flush().await?;
+        let response = TcpClPacket::parse(&mut self.stream).await?;
+        debug!("Received session parameters");
+        if let TcpClPacket::SessInit(mut data) = response {
+            let keepalive = sess_init_data.keepalive.min(data.keepalive);
+            sess_init_data.keepalive = keepalive;
+            data.keepalive = keepalive;
+            Ok((sess_init_data, data))
+        } else {
+            Err(TcpClError::UnexpectedPacket.into())
+        }
+    }
+
+    async fn exchange_contact_header(&mut self) -> anyhow::Result<ContactHeaderFlags> {
+        self.send_contact_header(ContactHeaderFlags::default())
+            .await?;
+        self.receive_contact_header().await
+    }
+
+    async fn send_contact_header(&mut self, flags: ContactHeaderFlags) -> anyhow::Result<()> {
+        self.stream.write(b"dtn!").await?;
+        self.stream.write_u8(4).await?;
+        self.stream.write_u8(flags.bits()).await?;
+        self.stream.flush().await?;
+        Ok(())
+    }
+
+    async fn receive_contact_header(&mut self) -> anyhow::Result<ContactHeaderFlags> {
+        let mut buf: [u8; 6] = [0; 6];
+        self.stream.read_exact(&mut buf).await?;
+        if &buf[0..4] != b"dtn!" {
+            bail!("Invalid magic");
+        }
+        if buf[4] != 4 {
+            bail!("Unsupported version");
+        }
+        Ok(ContactHeaderFlags::from_bits_truncate(buf[5]))
+    }
+
+    async fn connect(
+        mut self,
+        sessions: &mut impl DerefMut<Target = HashMap<SocketAddr, Sender<(SocketAddr, ByteBuffer)>>>,
+    ) {
+        // Phase 1
+        debug!("Exchanging contact header");
```

Comment:
as multiple tcpcl connections could start in parallel, always add socketaddr to log output for debugging purposes. at some point log level info should also log the connection process but only once